Functionality to filter/subset PPA full-text corpus by source id

rlskoeser opened the pull request:
ref #5

laurejt commented on src/corppa/utils/filter.py:
```diff
@@ -0,0 +1,83 @@
+    """
+    # read the id file and generate a list of ids
+    with open(idfile) as idfile_content:
+        source_ids = [line.strip() for line in idfile_content.readlines()]
```
```suggestion
        source_ids = [line.strip() for line in idfile_content]
```
```

laurejt commented on src/corppa/utils/filter.py:
```diff
@@ -0,0 +1,83 @@
+        help="path to PPA full-text corpus to be "
+        + "filtered; must be a JSONL file (compressed or not)",
+    )
+    parser.add_argument(
+        "output", help="filename where the filtered corpus should be saved"
+    )
```
What kind of behavior is desired for missing/inappropriate file extensions? The script will work, producing a json file. Perhaps ".json" should be appended to the filename in these cases?

rlskoeser commented on src/corppa/utils/filter.py:
I think I had that at one point and then I took it out! I wasn't sure if we want both ways to run it - if you install with pip / pyproject.toml it can be run with the name defined there.   But it doesn't hurt to have options; I'll add this back in.

rlskoeser commented:
@laurejt I pushed another set of changes with a bit more documentation and error handling for the script/main function.  I added type hints in a few places because I'd like to use a python docs solution that uses them.  I switched the parameter order - when I was demoing the script, it seemed more logical to me that the inputs should come before the desired output filename.

This is ready for you to test now. Would it be useful for me to draft a testing checklist on #5 ?

rlskoeser commented:
ha, belay that 😅 - messed myself up with the argument order change, adding a bit more error checking

rlskoeser commented:
OK! Now updated with better error handling on the files and corresponding updated unit tests.  

I started a testing checklist on https://github.com/Princeton-CDH/ppa-nlp/issues/5

laurejt commented on src/corppa/utils/filter.py:
```diff
@@ -97,13 +123,33 @@ def main():
+
     # if requested output filename has no extension, add jsonl
+    output_filename = args.output
     if os.path.splitext(output_filename)[1] == "":
```
This is assuming an "extension" simply corresponds to the existence of a `.` in the filename. This is probably fine, but will produce `.jsonl` files without the appropriate extension in some cases.

rlskoeser replied:
I don't think we should worry about that now as it seems unlikely, but if you have specific cases in mind we could document in case we want to revisit.

laurejt commented on src/corppa/utils/filter.py:
```diff
@@ -97,13 +123,33 @@ def main():
+        # catch known possible errors and display briefly
+        # with the type of error and the brief message
+        print(f"{err.__class__.__name__}: {err}")
+        exit(-1)
```
See comment for Line 128

```suggestion
        sys.exit(-1)
```

rlskoeser commented on src/corppa/utils/filter.py:
```diff
@@ -97,13 +123,33 @@ def main():
-    output_filename = args.output
+    if not os.path.exists(args.idfile):
+        print(f"Error: idfile {args.idfile} does not exist")
+        exit(-1)
```
I did not know that, thank you for flagging and sharing the link!

rlskoeser commented on src/corppa/utils/filter.py:
```diff
@@ -97,13 +123,33 @@ def main():
+    if not os.path.exists(args.idfile):
+        print(f"Error: idfile {args.idfile} does not exist")
+        exit(-1)
+    elif os.path.getsize(args.idfile) == 0:
```
yes, true - I thought this was reasonable for a simple check before processing (this was after I accidentally obliterated my id text file and replaced it with an empty file 😂 )

rlskoeser commented:
@laurejt thanks for the corrections and improvements. sys.exit is also much nicer to test, fwiw (I'm now using `pytest.raises(SystemExit)`). I've incorporated your changes and updated the unit tests.
